File: python/bob/bob.py
def hey(phrase):
    if silence(phrase):
        return "Fine. Be that way!"
    elif shouted(phrase) and question(phrase):
        return "Calm down, I know what I'm doing!"
    elif shouted(phrase):
        return "Whoa, chill out!"
    elif question(phrase):
        return "Sure."
    else:
        return "Whatever."


def silence(phrase):
    return phrase == "" or phrase.isspace()


def shouted(phrase):
    return phrase.strip().isupper()


def question(phrase):
    return phrase.strip().endswith("?")

File: javascript/bob/bob.js
class Bob {
  hey(message) {
    if (Bob.getsTheSilentTreatment(message)) {
      return 'Fine. Be that way!';
    }

    if (this.hasAtLeastOneLetter(message) && Bob.isYelledAt(message) && Bob.isQuestioned(message)) {
      return 'Calm down, I know what I\'m doing!';
    }

    if (this.hasAtLeastOneLetter(message) && Bob.isYelledAt(message)) {
      return 'Whoa, chill out!';
    }

    if (Bob.isQuestioned(message)) {
      return 'Sure.';
    }

    return 'Whatever.';
  }

  static isYelledAt(message) {
    return message === message.toUpperCase(); // message is all caps
  }

  static isQuestioned(message) {
    return /\?$/.test(message); // message ends with a question
  }

  static getsTheSilentTreatment(message) {
    return !message.trim(); // message is not empty
  }

  hasAtLeastOneLetter(message) {
    return /[a-zA-Z]/.test(message);
  }

}

export default Bob;
